improver/calibration/reliability_calibration.py

```diff
@@ -0,0 +1,414 @@
+
+        Args:
+            forecast (numpy.ndarray):
+                An array containing data over an xy slice for a single validity
+                time and threshold.
+            truth (numpy.ndarray):
+                An array containing a thresholded gridded truth at an
+                equivalent validity time to the forecast array.
+        Returns:
+            numpy.ndarray:
+                An array containing reliability table data for a single time
+                and threshold.
+        """
+        observation_counts = []
+        forecast_probabilities = []
+        forecast_counts = []
+
+        for bin_min, bin_max in self.probability_bins:
+
+            observation_mask = (((forecast >= bin_min) & (forecast <= bin_max))
+                                & (np.isclose(truth, 1))).astype(int)
+            forecast_mask = ((forecast >= bin_min) &
+                             (forecast <= bin_max)).astype(int)
+            forecasts_probability_values = forecast * forecast_mask
+
+            observation_counts.append(observation_mask)
+            forecast_probabilities.append(forecasts_probability_values)
+            forecast_counts.append(forecast_mask)
+
+        reliability_table = np.stack([
+                                np.stack(observation_counts),
+                                np.stack(forecast_probabilities),
+                                np.stack(forecast_counts)])
+        return reliability_table.astype(np.float32)
+
+    def process(self, historic_forecasts, truths):
+        """
+        Slice data over threshold and time coordinates to construct reliability
+        tables. These are summed over time to give a single table for each
+        threshold, constructed from all the provided historic forecasts and
+        truths.
+
+        Args:
+            historic_forecasts (iris.cube.Cube):
+                A cube containing the historical forecasts used in calibration.
+                These are expected to all have a consistent cycle hour, that is
+                the hour in the forecast reference time.
+            truths (iris.cube.Cube):
+                A cube containing the thresholded gridded truths used in
+                calibration.
+        Returns:
+            iris.cube.CubeList:
+                A cubelist of reliability table cubes, one for each threshold
+                in the historic forecast cubes.
+        """
+        historic_forecasts, truths = filter_non_matching_cubes(
+            historic_forecasts, truths)
+
+        threshold_coord = find_threshold_coordinate(historic_forecasts)
+        time_coord = historic_forecasts.coord('time')
+
+        self._check_forecast_consistency(historic_forecasts)
+        reliability_cube = self._create_reliability_table_cube(
+            historic_forecasts, threshold_coord)
+
+        reliability_tables = iris.cube.CubeList()
+        threshold_slices = zip(historic_forecasts.slices_over(threshold_coord),
```

Comment:
I was wondering if this fails if there are mismatched threshold coordinates on the truth and the forecasts. I guess this is the point it will fail.

Reply:
Zip doesn't actually fail if the two mismatch in length, it will create an object with N elements where N is the length of the shorter of the two inputs. This means it could be comparing unmatching thresholds with no warning.

I've added a test into the process method. It remains to be seen whether this is now too stringent; i.e. should we just check the points match rather than the entire coordinate in case there is a random attribute on one threshold coordinate and not the other?

Reply:
At the moment I don't have a problem with it matching the coordinates rather than just the points.